src/SingleSignatureDirtyHack.sol

```diff
@@ -269,4 +314,90 @@ abstract contract SingleSignatureDirtyHack is IERC5267, AbstractContext {
         transferParams.token.safeTransferFrom(transferParams.from, transferParams.to, transferParams.requestedAmount);
         return true;
     }
+
+    function transferFromPermit(
+        string calldata typeSuffix,
+        TransferParams calldata transferParams,
+        PackedSignature calldata sig
+    ) external returns (bool) {
+        _checkDeadline(transferParams.deadline);
+
+        bytes32 structHash = _hashStruct(typeSuffix, transferParams);
+        bytes32 signingHash = _encodePermitHash(
+            transferParams.token, transferParams.from, transferParams.nonce, transferParams.amount, structHash
+        );
+        
+        address signer;
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+            let vs := calldataload(add(0x20, sig))
+            let owner := calldataload(add(0x60, transferParams)) // from
+            let value := calldataload(add(0xa0, transferParams)) // amount
+
+            // get signer
+            mstore(add(0xd4, ptr), and(vs, 0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)) // s
+            mstore(add(0xb4, ptr), calldataload(sig)) // r
+            mstore(add(0x94, ptr), add(0x1b, shr(0xff, vs))) // v
+            mstore(add(0x74, ptr), signingHash)
+            
+            pop(staticcall(gas(), 0x01, add(0x74, ptr), 0x80, 0x00, 0x20))
+            signer := mul(mload(0x00), eq(returndatasize(), 0x20))
+
+            // keep adding parameters to call permit
+            // TODO: check that calling ecrecover doesn't affect the previously added parameters
```

Comment:
`mload(0x40)` can never be below `0x80`, so specifying `0x00, 0x20` as the returndata region for `ecrecover` can't clobber any of the values that you put after the free memory pointer